src/execution_plan/ops/op_value_hash_join.c

```diff
@@ -0,0 +1,289 @@
+    return OP_OK;
+}
+
+/* Produce a record by joining 
+ * records coming from the left and right hand side 
+ * of this operation. */
+Record ValueHashJoinConsume(OpBase *opBase) {
+    OpValueHashJoin *op = (OpValueHashJoin*)opBase;
+    OpBase *right_child = op->op.children[1];
+
+    // Eager, pull from left branch until depleted.
+    if(op->cached_records == NULL) {
+        _cache_records(op);
+        // Sort cache on intersect node ID.
+        _sort_cached_records(op);
+    }
+
+    /* Try to produce a record:
+     * given a right hand side record R,
+     * evaluate V = exp on R,
+     * see if there are any cached records
+     * which V evaluated to V:
+     * X in cached_records and X[idx] = V
+     * return merged record:
+     * X merged with R. */
+
+    Record l;
+    if(op->number_of_intersections > 0) {
+        while((l = _get_intersecting_record(op))) {
+            Record_Merge(&op->rhs_rec, l);
+            return Record_Clone(op->rhs_rec);            
+        }
+    }
+
+    /* If we're here there are no more
+     * left hand side records which intersect with R
+     * discard R. */
+    if(op->rhs_rec) Record_Free(op->rhs_rec);
+
+    /* Try to get new right hand side record
+     * which intersect with a left hand side record. */
+    while(true) {
+        // Pull from right branch.
+        op->rhs_rec = right_child->consume(right_child);
+        if(!op->rhs_rec) return NULL;
+
+        // Get value on which we're intersecting.
+        SIValue v = AR_EXP_Evaluate(op->rhs_exp, op->rhs_rec);
+
+        // No intersection, discard R.
+        if(!_set_intersection_idx(op, v)) {
+            Record_Free(op->rhs_rec);
+            continue;
+        }
+
+        // Found atleast one intersecting record.
+        l = _get_intersecting_record(op);
+        Record_Merge(&op->rhs_rec, l);
+        return Record_Clone(op->rhs_rec);
+    }
+}
+
+OpResult ValueHashJoinReset(OpBase *ctx) {
+    OpValueHashJoin *op = (OpValueHashJoin*)ctx;
+    op->intersect_idx = -1;
+    op->number_of_intersections = 0;
+
+    // Clear cached records.
+    if(op->rhs_rec) {
+        Record_Free(op->rhs_rec);
+        op->rhs_rec = NULL;
+    }
+
+    if(op->cached_records) {
+        uint record_count = array_len(op->cached_records);
+        for(uint i = 0; i < record_count; i++) {
+            Record r = op->cached_records[i];
+            Record_Free(r);
+        }
+        array_free(op->cached_records);
+        op->cached_records = NULL;
+    }
+
+    return OP_OK;
+}
+
+/* Frees valueHashJoin */
+void ValueHashJoinFree(OpBase *ctx) {
+    OpValueHashJoin *op = (OpValueHashJoin*)ctx;
+    // Free cached records.
+    if(op->rhs_rec) Record_Free(op->rhs_rec);
+
+    if(op->cached_records) {
+        uint record_count = array_len(op->cached_records);
+        for(uint i = 0; i < record_count; i++) {
+            Record r = op->cached_records[i];
+            Record_Free(r);
+        }
+        array_free(op->cached_records);
+    }
```

Comment:
Do we need to free `lhs_exp` and `rhs_exp`?

Reply:
Indeed we do now.